CARDS-1501: Upon deletion of subject and its questionnaires, refresh the questionnaires grid on the dashboard

veronikaslc opened the pull request.

veronikaslc committed:
CARDS-1501: Upon deletion of subject and its questionnaires, refresh the questionnaires grid on the dashboard

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/229209055-ccf973f5-cec1-45a3-870a-262fd693148a.png)

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -82,6 +82,18 @@ function LiveTable(props) {
 
   const globalLoginDisplay = useContext(GlobalLoginContext);
 
+  // When data is changed, trigger a new fetch in Forms table
+  useEffect(() => {
+    if (entryType == "Form") {
+      // subscribe event
+      window.addEventListener("SubjectDeleted",  refresh);
+      return () => {
+        // unsubscribe event
+        document.removeEventListener("SubjectDeleted",  refresh);
+      };
+    }
+  }, [entryType]);
+
```
Is there someplace else we can move this event subscription? The livetable should be a generic component, not concerned with logic for very specific types of data.

My first suggestion would be to observe the SubjectDeleted event in FormView and trigger an update for the Forms livetable from there. Other suggestions that preserve a focused functionality of our generic components are very welcome.

veronikaslc replied:
refactored in [0ef9958](https://github.com/data-team-uhn/cards/pull/1434/commits/0ef99584bd013067d76c8c761b9a8d0dc125fec5)

veronikaslc committed:
CARDS-1501: Upon deletion of subject and its questionnaires, refresh the questionnaires grid on the dashboard
Relocated the subject deleted event listener to the FormView
Enhanced LiveTable actions with the onActionDone callback param to perform when action is done (in this case we trigger the event dispatch)

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -96,6 +96,11 @@ function LiveTable(props) {
     }
   }, [customUrl]);
 
+  // Initialize the component: if there's no data loaded yet, fetch the first page
+  useEffect(() => {
+    if (fetchStatus.currentRequestNumber == -1) fetchData(paginationData, true);
+  }, [fetchStatus.currentRequestNumber]);
+
```
why this move?

veronikaslc replied:
codestyle for better development experience while working on this component

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -225,12 +230,14 @@ function LiveTable(props) {
   let makeActions = (entry, actions, index) => {
-    let content = actions.map((Action, index) => {
+    let content = actions.map((item, index) => {
+      let Action = item.component;
```
can we  avoid some changes to other livetables unrelated to this PR, such as Questionnaires.jsx, by allowing `actions` entries to be either a component or a json with `component` and `onActionDone`, and doing  something like
```
let Action = item.component ?? item;
```
here instead?

veronikaslc replied:
That could bring more confusion later. Imho keeping the `actions` type uniform is a cleaner approach and helps devs in case of future needs to add more. Also the changes are minor.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/LiveTable.jsx:
```diff
@@ -225,12 +230,14 @@ function LiveTable(props) {
         entryPath={entry["@path"]}
         entryName={getEntityIdentifier(entry)}
-        onComplete={refresh}
+        onComplete={() => {refresh(); onDone && onDone();}}
```
We can make it even shorter:
```suggestion
        onComplete={() => {refresh(); onDone?.();}}
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/FormView.jsx:
```diff
@@ -50,6 +50,8 @@ function FormView(props) {
   const [ qFetchSent, setQFetchStatus ] = useState(false);
   const [ filtersJsonString, setFiltersJsonString ] = useState(new URLSearchParams(window.location.hash.substring(1)).get("forms:filters"));
+  // When a new subject is added, state will be updated from event listener and trigger a livetable refresh
+  const [ requestFetchData, setRequestFetchData ] = useState();
```
I would name it `requestForRefresh` instead

veronikaslc committed:
CARDS-1501: Upon deletion of subject and its questionnaires, refresh the questionnaires grid on the dashboard
Better names for the params;
Some codestyle improvement

veronikaslc committed:
CARDS-1501: Upon deletion of subject and its questionnaires, refresh the questionnaires grid on the dashboard
Refactored the code: introduced an extension of DeleteBetton that dispatches the refresh event for a lifetable

veronikaslc committed:
[misc] Code cleanup

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/SubjectView.jsx:
```diff
@@ -38,7 +38,7 @@ import withStyles from '@mui/styles/withStyles';
 import AssignmentIndIcon from '@mui/icons-material/AssignmentInd';
 import LaunchIcon from '@mui/icons-material/Launch';
-import DeleteButton from "./DeleteButton.jsx";
+import DeleteRefreshButton from "./DeleteRefreshButton.jsx";
```
This file has not been committed. Also, please rename to `DeleteWithRefresh`.

veronikaslc committed:
CARDS-1501: Upon deletion of subject and its questionnaires, refresh the questionnaires grid on the dashboard
Renamed delete with refresh button

sdumitriu reviewed (commented):
Code is nice and clean so far. Should we also use the refreshing delete button on the form view as well?

marta- commented:
> Code is nice and clean so far. Should we also use the refreshing delete button on the form view as well?

As of now the only scenario I'm aware of where deleting a form may trigger changes in subject tables would be if we used CopyAnswers to augment subject jsons with some data from the forms, and displayed those values in the table. We don't have this setup in practice (in the patient portal's Clinic Dashboard we do rely on CopyAnswers, but that dashboard has no action buttons). I vote for not using the refreshing delete button in FormView just yet,, because on the User Dashboard it will trigger calls to the server with no visible benefits.

marta- reviewed (approved):
Tested & reviewed, looks great

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (approved):
One minor thing bugging me: when the tables are refreshed, they always go back to the first page.
